shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java

```diff
@@ -215,12 +273,43 @@ public void handleMessageFromDart(
             Trace.endSection();
           }
         };
-    @NonNull
     final DartMessengerTaskQueue nonnullTaskQueue =
         taskQueue == null ? platformTaskQueue : taskQueue;
     nonnullTaskQueue.dispatch(myRunnable);
   }
 
+  @Override
+  public void handleMessageFromDart(
+      @NonNull String channel, @Nullable ByteBuffer message, int replyId, long messageData) {
+    // Called from the ui thread.
+    Log.v(TAG, "Received message from Dart over channel '" + channel + "'");
+
+    HandlerInfo handlerInfo;
+    boolean messageDeferred;
+    synchronized (handlersLock) {
+      handlerInfo = messageHandlers.get(channel);
+      messageDeferred = (enableBufferingIncomingMessages && handlerInfo == null);
+      if (messageDeferred) {
+        // The channel is not defined when the Dart VM sends a message before the channels are
+        // registered.
+        //
+        // This is possible if the Dart VM starts before channel registration, and if the thread
+        // that registers the channels is busy or slow at registering the channel handlers.
+        //
+        // In such cases, the task dispatchers are queued, and processed when the channel is
+        // defined.
+        if (!bufferedMessages.containsKey(channel)) {
+          bufferedMessages.put(channel, new LinkedList<>());
```

Comment:
When I implemented Channel Buffers I got a lot of flak about having unbounded buffering of platform channels.  The resolution we ultimately decided upon was a RingBuffer of size 1 with control messages that can be used to adjust the size.  The concern was that incorrect usage could blow up memory.

I thought it was overly cautious.  Maybe we should have a RingBuffer at least so it isn't completely unbounded.

Reply:
is the suggestion to override messages? How is that desired in an app that runs into this situation?

Reply:
A RingBuffer drops messages after its capacity is exhausted, dropping messages isn't a desired behavior but also queueing up infinite messages isn't desired.

Messages that write data can have 2 different semantics, `x = y` or `x = f(x, y)`.  In the first case dropping messages isn't actually an error.  The only thing that matters is the last operation.  In the latter case, it is a problem.  It impossible to know which write semantics the channel has without the user declaring it.

So, in summary, a RingBuffer doesn't always drop messages and when it does it isn't necessarily a logical error.  If I were king of Flutter I'd say just chose some reasonable upper bounds for the RingBuffer, like 1024.  I can tell you though that we've had this discussion before as a team and the resolution was that the RingBuffer size should be 1.

attn @Hixie who was involved in deciding how Channel Buffers should work.

Reply:
If a RingBuffer is preferred, I'm happy to change it. I wasn't aware that it's ok to drop some messages. e.g. could that yield to invalid state in a way that is really hard to debug?

Reply:
Is the plan that buffering messages only happens for a limited amount of time via `enableBufferingIncomingMessages`?  It doesn't appear to be doing that but if that's the case it isn't as bad since you only have a limited amount of time to queue up infinite messages.

What is the plan for messages that are sent, buffered, but no handler is ever registered for them?  Are you going to hold onto them indefinitely?  In the past they would have gotten an exception that maybe was ignored but now will be held onto.  Maybe once we turn off the time frame where messages are buffered all of those messages without handlers could be cleared out to match the previous behavior.

Reply:
> Is the plan that buffering messages only happens for a limited amount of time via enableBufferingIncomingMessages?

right on

> What is the plan for messages that are sent, buffered, but no handler is ever registered for them?

Added some logic to address this in disableBufferingIncomingMessages